WIP first pass on ActivityPub knex repositories

aileen opened the pull request:
ref MOM-123

This is 100% a WIP and not ready yet. Many questions need to be answered first before proceeding

---

🚧 **WIP** 🚧

aileen committed:
Added table for `activitypub`

ref MOM-122

- Added prototype table for `activitypub`
- The table is classed as prototype, as we're not expecting this to be the final version

aileen commented on ghost/ghost/src/core/activitypub/actor.entity.ts:
```diff
@@ -16,7 +16,9 @@ type ActorData = {
     inbox: Activity[];
     following: {id: URI, username?: string;}[];
-    followers: {id: URI;}[];
+    followers: {id: URI;}[],
```
@allouis I wasn't sure if we need `followers` here. There's some funky logic with `activities` that I haven't fully gotten yet, so maybe it's wrong to add it here?

aileen commented on ghost/ghost/src/db/knex/activity.repository.knex.ts:
```diff
@@ -0,0 +1,53 @@
+        @Inject('knex') private readonly knex: any
+    ) {}
+
+    #modelToActivity(model: any) {
```
@allouis I don't know if that's needed, but I expect that the repository has to return an `Actor` type, so not sure how else to serialise it...

allouis replied:
The Repository deffo has to return an `Activity` type - this should be giving TS errors 🤔 

You should be able to `return new Activity(data)` here, or if you need a factory, create a new factory method and do `return Activity.create(data)`

aileen commented on ghost/ghost/src/db/knex/actor.repository.knex.ts:
```diff
@@ -0,0 +1,148 @@
+        @Inject('SettingsCache') settingsCache: SettingsCache,
+        @Inject('DomainEvents') domainEvents: DomainEvents
+    ) {
+        this.#defaultActor = Actor.create({
```
Same as above, but I was able to `create` a new actor here. Again, not sure if that is the right approach.

aileen replied:
Also not sure, if we need a default actor, or if this is really the right place to create them?

allouis replied:
Fixtures is the correct place for defaults, but here is good for now, we can easily make changes to the values in dev

allouis replied:
Although - this is going to blow away the following list each time? I think need to be careful about that

aileen commented on ghost/ghost/src/db/knex/actor.repository.knex.ts:
```diff
@@ -0,0 +1,148 @@
+        }
+    }
+
+    async save(actor: Actor) {
```
@allouis I tried adding a follow via the frontend, and I can see that the workflow for this is via adding it as an activity first, and then only the `acceptFollow` actually add the follow to the array of the entity. But this method is only called from the `postToInbox` method, which I never managed to actually call. I think, I need some more explanation here about how it's expected to work.

allouis replied:
So the flow is like:

1. Alice creates Follow Activity with Bob as the `object` and sends it to Bob's Inbox
2. Bob receives the Follow Activity, and creates an Accept Activity with the Follow Activity as the `object` and sends it to Alice's Inbox 

So when you're testing follow, you need to have ngrok setup, so that the site you're following is able to send the Accept back to you, which will then get handled in the `postToInbox` method.

allouis commented on ghost/ghost/src/core/activitypub/actor.entity.ts:
```diff
@@ -84,6 +86,18 @@ export class Actor extends Entity<ActorData> {
+        return this.attr.featured;
+    }
+
+    get privateKey() {
```
I don't think we want/need to expose this

allouis commented on ghost/ghost/src/db/knex/activity.repository.knex.ts:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    async save(activity: Activity) {
+        const date = this.knex.raw('CURRENT_TIMESTAMP');
```
We should be reading this from the entity, rather than creating it ourself

allouis reviewed (commented):
Few comments!

aileen committed:
WIP first pass on actor repository

ref MOM-123

This is 100% a WIP and not ready yet. Many questions need to be answered first before proceeding

aileen committed:
Added types for DB data

ref MOM-123

aileen committed:
WIP first pass on activity repository

ref MOM-123

- Added a create factory to the `activity` entity, so we can return a new `Activity` entity when fetching the data from the db

aileen committed:
Added timestamps to entities

ref MOM-123

Instead of create a knex timestamp, we take the one that exists on the entity already.
For this, I added a new date validation to the `create` factories, which ensures that the entities always have the correct timestamp

aileen committed:
Fixed linting error

aileen committed:
Fixed activities being stored in DB as entity and other small things

aileen committed:
Switched to use knex repositories

aileen commented:
Re-opened against `activitypub` branch → https://github.com/TryGhost/Ghost/pull/20239
